genomedata/__init__.py

```diff
@@ -494,6 +453,131 @@ def vars(self):
+        for filepath in sorted(self.filepath.files("*" + SUFFIX)):
+            # pass through __getitem__() to allow memoization
+            yield self[filepath.stem]
+
+    def __getitem__(self, name):
+        # If we have opened the chromosome already
+        res = super().__getitem__(name)
+        # Otherwise
+        if not res:
+            # Return a new chromosome from from hdf5 file
+            res = _HDF5DirectoryChromosome(
+                  self.filepath.joinpath(name + SUFFIX),
+                  *self.args, **self.kwargs)
+            self.open_chromosomes[name] = res
+
+        return res
+
+    def tracknames_continuous(self):
+        # check that all chromosomes have the same tracknames_continuous
+        res = None
+        for chromosome in self:
+            if res is None:
+                res = chromosome.tracknames_continuous
+            else:
+                assert res == chromosome.tracknames_continuous
+
+        return res
+
+    def _add_trackname(self, trackname):
+        for chromosome in self:
+            chromosome._add_trackname(trackname)
+
+    def create(self, name):
+        return self[name]
+
+    def close(self):
+        # Close all the chromosomes so they know they shouldn't be read
+        for chrom in self:
+            chrom.close()
+
+        super().close()
+
+
+class _HDF5SingleFileChromosomeList(_ChromosomeList):
+    def __init__(self, filepath, *args, **kwargs):
+        self.h5file = _open_file(filepath, *args, **kwargs)
+        self._file_attrs = self.h5file.root._v_attrs
+
+        super().__init__()
+
+    def __iter__(self):
+        # Iterate over child group of root
+        for group in self.h5file.iter_nodes("/", classname="Group"):
+            groupname = group._v_name
+            # Pass through to __getitem__()
+            yield self[groupname]
+
+    def __getitem__(self, name):
+        # If we have opened the chromosome already
+        res = super().__getitem__(name)
+        # Otherwise
+        if not res:
+            # Return a new chromosome from from hdf5 file
+            res = _HDF5FileChromosome(self.h5file, where="/" + name)
+            self.open_chromosomes[name] = res
+
+        return res
+
+    def _add_trackname(self, trackname):
+        _hdf5_add_trackname(self.h5file, trackname)
+
+    def tracknames_continuous(self):
+        # Tracknames are stored at the root of each file, so we can
+        # access them directly in this case
+        return [decode_trackname(trackname)
+                for trackname in self._file_attrs.tracknames]
+
+    def create(self, name):
+        self.h5file.create_group("/", name, filters=FILTERS_GZIP)
+        return self[name]
+
+    def close(self):
+        # Close all the chromosomes so they know they shouldn't be read. Do
+        # this before closing h5file in case the chromosomes need access
+        # to it in closing.
+        for chrom in self:
+            chrom.close()
```

Comment:
Is this going to try to close `Chromosome`s that haven't been opened?

Reply:
No since each chromosome is iterated through `__getitem__` which does the sanity checking of only giving open/valid chromosomes

Reply:
Can you add a comment to that effect?